### PR TITLE
MINIFICPP-1759 Fix linux/libc++ issues, use libc++ on ubuntu

### PR DESCRIPTION
**Pull request by szaszm**
- not fixed issue: OpenCV refuses to compile with libc++/C++20
- ~Bustache no longer requires boost.~ As it turns out, libboostache depends on boost. OpenCV does, too. It's also required for some test on windows. I touched a few boost-related places.

---

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/bustache/ApplyTemplate.cpp [hunk 1]
```diff
@@ -19,81 +19,76 @@
...
 }
 
-ApplyTemplate::WriteCallback::WriteCallback(const std::string &path, const std::shared_ptr<core::FlowFile> &flow_file) {
-  template_file_ = path;
-  flow_file_ = flow_file;
-}
-
-int64_t ApplyTemplate::WriteCallback::process(const std::shared_ptr<io::BaseStream>& stream) {
-  logger_->log_info("ApplyTemplate reading template file from %s", template_file_);
-  boost::iostreams::mapped_file_source file(template_file_);
```
Comment (szaszm) [inline extensions/bustache/ApplyTemplate.cpp:76]: This line was crashing (segfault) when compiled with libc++, so I removed it. I also saw a bunch of refactoring opportunities, so ApplyTemplate has quite a few changes. As it turns out, this line was the only reason the bustache extension required boost, so I removed that requirement and checked a few boost-related pieces of code.

File: extensions/expression-language/Expression.cpp [hunk 2]
```diff
@@ -636,7 +636,7 @@ Value expr_toDate(const std::vector<Value> &args) {
   auto utc = date::locate_zone("UTC");
   auto utct = date::make_zoned(utc, t);
   auto zt = date::make_zoned(zone, utct.get_local_time());
-  return Value(std::chrono::duration_cast<std::chrono::milliseconds>(zt.get_sys_time().time_since_epoch()).count());
+  return Value(int64_t{std::chrono::duration_cast<std::chrono::milliseconds>(zt.get_sys_time().time_since_epoch()).count()});
```
Comment (szaszm) [inline extensions/expression-language/Expression.cpp:639]: With libc++, the count return value is `long long`, which is not convertible to `int64_t` (alias to `long`). `Value` doesn't have a `long long` constructor, so I had to add a non-narrowing cast to these places.

File: extensions/standard-processors/processors/GetFile.cpp [hunk 3]
```diff
@@ -133,7 +130,9 @@ void GetFile::onSchedule(core::ProcessContext *context, core::ProcessSessionFact
     core::Property::StringToInt(value, request_.minSize);
   }
 
-  context->getProperty(PollInterval.getName(), request_.pollInterval);
+  context->getProperty<core::TimePeriodValue>(PollInterval)
+      | utils::map(&core::TimePeriodValue::getMilliseconds)
+      | utils::map([this](std::chrono::milliseconds ms) { request_.pollInterval = ms; });
```
Comment (szaszm) [inline extensions/standard-processors/processors/GetFile.cpp:135]:
```
In file included from /home/szaszm/nifi-minifi-cpp-3/extensions/standard-processors/processors/GetFile.cpp:18:
In file included from /home/szaszm/nifi-minifi-cpp-3/extensions/standard-processors/processors/GetFile.h:29:
In file included from /home/szaszm/nifi-minifi-cpp-3/extensions/standard-processors/../../libminifi/include/core/Processor.h:39:
/home/szaszm/nifi-minifi-cpp-3/extensions/standard-processors/../../libminifi/include/core/ConfigurableComponent.h:237:15: error: no matching conversion for static_cast from 'const org::apache::nifi::minifi::core::PropertyValue' to 'std::chrono::duration<long long, std::ratio<1, 1000>>'
      value = static_cast<T>(property.getValue());  // cast throws if the value is invalid
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/szaszm/nifi-minifi-cpp-3/extensions/standard-processors/../../libminifi/include/core/ProcessorNode.h:73:30: note: in instantiation of function template specialization 'org::apache::nifi::minifi::core::ConfigurableComponent::getProperty<std::chrono::duration<long long, std::ratio<1, 1000>>>' requested here
      return processor_cast->getProperty<T>(name, value);
                             ^
/home/szaszm/nifi-minifi-cpp-3/extensions/standard-processors/../../libminifi/include/core/ProcessContext.h:386:29: note: in instantiation of function template specialization 'org::apache::nifi::minifi::core::ProcessorNode::getProperty<std::chrono::duration<long long, std::ratio<1, 1000>>>' requested here
    return processor_node_->getProperty<typename std::common_type<T>::type>(name, value);
                            ^
/home/szaszm/nifi-minifi-cpp-3/extensions/standard-processors/../../libminifi/include/core/ProcessContext.h:118:12: note: in instantiation of function template specialization 'org::apache::nifi::minifi::core::ProcessContext::getPropertyImp<std::chrono::duration<long long, std::ratio<1, 1000>>>' requested here
    return getPropertyImp<typename std::common_type<T>::type>(name, value);
           ^
/home/szaszm/nifi-minifi-cpp-3/extensions/standard-processors/processors/GetFile.cpp:136:12: note: in instantiation of function template specialization 'org::apache::nifi::minifi::core::ProcessContext::getProperty<std::chrono::duration<long long, std::ratio<1, 1000>>>' requested here
  context->getProperty(PollInterval.getName(), request_.pollInterval);
           ^
/usr/include/c++/v1/chrono:1023:28: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const org::apache::nifi::minifi::core::PropertyValue' to 'const std::chrono::duration<long long, std::ratio<1, 1000>>' for 1st argument
class _LIBCPP_TEMPLATE_VIS duration
                           ^
/usr/include/c++/v1/chrono:1023:28: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const org::apache::nifi::minifi::core::PropertyValue' to 'std::chrono::duration<long long, std::ratio<1, 1000>>' for 1st argument
/usr/include/c++/v1/chrono:1075:18: note: candidate template ignored: requirement 'is_convertible<org::apache::nifi::minifi::core::PropertyValue, long long>::value' was not satisfied [with _Rep2 = org::apache::nifi::minifi::core::PropertyValue]
        explicit duration(const _Rep2& __r,
                 ^
/usr/include/c++/v1/chrono:1087:9: note: candidate template ignored: could not match 'duration<type-parameter-0-0, type-parameter-0-1>' against 'const org::apache::nifi::minifi::core::PropertyValue'
        duration(const duration<_Rep2, _Period2>& __d,
        ^
/usr/include/c++/v1/chrono:1068:9: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
        duration() = default;
        ^
```

File: libminifi/include/utils/OptionalUtils.h [hunk 4]
```diff
@@ -43,22 +43,19 @@ struct is_optional<std::optional<T>, void> : std::true_type {};
 namespace detail {
 // map implementation
 template<typename SourceType, typename F>
-auto operator|(const std::optional<SourceType>& o, map_wrapper<F> f) noexcept(noexcept(std::invoke(std::forward<F>(f.function), *o)))
-    -> std::optional<typename std::decay<decltype(std::invoke(std::forward<F>(f.function), *o))>::type> {
-  if (o.has_value()) {
-    return std::make_optional(std::invoke(std::forward<F>(f.function), *o));
+auto operator|(std::optional<SourceType> o, map_wrapper<F> f) noexcept(noexcept(std::invoke(std::forward<F>(f.function), *std::move(o)))) {
+  using cb_result = std::decay_t<std::invoke_result_t<F, SourceType>>;
+  if constexpr(std::is_same_v<cb_result, void>) {
```
Comment (szaszm) [inline libminifi/include/utils/OptionalUtils.h:48]: Added support for `void` return types from the callback, in which case `map` just returns `void`. There is no such thing as `optional<void>`. This simplified the change in `GetFile`.

File: extensions/bustache/ApplyTemplate.cpp [hunk 5]
```diff
@@ -19,81 +19,76 @@
...
+namespace org::apache::nifi::minifi::processors {
+
+const core::Property ApplyTemplate::Template("Template", "Path to the input mustache template file", "");
+const core::Relationship ApplyTemplate::Success("success", "success operational on the flow record");
+
+namespace {
+class WriteCallback : public OutputStreamCallback {
+ public:
+  WriteCallback(std::filesystem::path templateFile, const core::FlowFile& flow_file)
+      :template_file_{std::move(templateFile)}, flow_file_{std::move(flow_file)}
```
Comment (adam-markovics) [inline extensions/bustache/ApplyTemplate.cpp:41]: std::move does not have an effect on flow_file

**Review by adam-markovics — APPROVED** [review]

**Review by lordgamez — APPROVED** [review]

File: extensions/standard-processors/processors/GetFile.cpp (hunk 3 above)
Comment (fgerlits) [inline extensions/standard-processors/processors/GetFile.cpp:135]: this is done elsewhere like this:
```suggestion
  if (const auto poll_interval = context->getProperty<core::TimePeriodValue>(PollInterval)) {
      request_.pollInterval = poll_interval->getMilliseconds();
  }
```
which I find more readable

Comment (szaszm) [reply extensions/standard-processors/processors/GetFile.cpp]: Incorporated your suggestion in b6abaeb

**Pull request closed without merging**